### PR TITLE
Loki: Fix timerange for query stats request

### LINKED ISSUE #69086: Loki: Stats endpoint for instant query should get time range based on duration
At the moment, the query size estimation is only (relatively) accurate when doing range queriers, because start and end of the stats query are defined by the values in the time picker.  However, when I do an instant query the time range is not relevant, but rather the range of the query (within the brackets) is. 

So if you have query `rate({a="B"}[5m])` and time range is `now` - `now-1h` , then for query estimation:
- for range query we would send the time range `now` - `now-1h`
= for instant query we would send the time range `now` - `now-5m`

### PR DESCRIPTION
**Pull request by itsgareth**
**What is this feature?**
currently the timerange we request for the query stats data is slightly incorrect.
this pr fixes the requested timerange.

**Why do we need this feature?**
more accurate response.

**Which issue(s) does this PR fix?**:
Fixes #69086

**Special notes for your reviewer:**

explanation of the changes below.
in the explanations i use relative time e.g. 'now-5m' however the actual request is made using unix time in NS

`Metric query` with `Type: Range`:
```
ExploreTimePicker: now to now-30m
Query: rate({label="value"} [5m])

Requested Timerange before:
{ start: 'now-30m', end: 'now'}

Requested Timerange after:
// we add the duration from the query, this accounts for the last datapoint
{ start: 'now-35m', end: 'now'}
```

`Metric query` with `Type: Instant`:
```
ExploreTimePicker: now to now-30m
Query: rate({label="value"} [5m])

Requested Timerange before:
{ start: 'now-30m', end: 'now'}

Requested Timerange after:
{ start: 'now-5m', end: 'now'}
```

`Logs query` with `Type: Range`:
```
NOTE: The behaviour for this didn't change.

ExploreTimePicker: now to now-30m
Query:{label="value"}

Requested Timerange before:
{ start: 'now-30m', end: 'now'}

Requested Timerange after:
{ start: 'now-30m', end: 'now'}
```

`Logs query` with `Type: Instant`:
```
You shouldn't use logs query with type instant, we now display:
"Query size estimate not available."
```

### CODE REVIEW

File: public/app/plugins/datasource/prometheus/querybuilder/shared/QueryOptionGroup.tsx [hunk 1]
```diff
@@ -10,16 +10,28 @@ import { QueryStats } from 'app/plugins/datasource/loki/types';
...
-  queryStats?: QueryStats | null;
+  queryStats?: QueryStats | string | null;
   children: React.ReactNode;
 }
 
 export function QueryOptionGroup({ title, children, collapsedInfo, queryStats }: Props) {
   const [isOpen, toggleOpen] = useToggle(false);
   const styles = useStyles2(getStyles);
 
+  const generateQueryStats = () => {
```
Comment (gabor) [inline public/app/plugins/datasource/prometheus/querybuilder/shared/QueryOptionGroup.tsx:21]: i recommend moving `generateQueryStats` and `convertUnits` outside of `QueryOptionsGroup`. they only use the `queryStats` value, so i think it's easier to think about them when they are simple functions that take the `queryStats` and return something (will be a bit faster too).

File: public/app/plugins/datasource/loki/components/stats.ts [hunk 2]
```diff
@@ -52,3 +57,39 @@ export function shouldUpdateStats(
...
+  const durationNodes = getNodesFromQuery(query.expr, [Duration]);
+  const durations = durationNodes.map((d) => query.expr.substring(d.from, d.to));
+
+  if (!isLogsQuery(query.expr)) {
+    if (query.queryType === LokiQueryType.Instant) {
+      // metric query with instant type
+      // we want the request timerange to be the query duration e.g. [5m]
+      // with the query -- rate({label="value"} [1h]) -- the range we want to request is: "now - 1h"
+      end = new Date().getTime() * NS_IN_MS;
+      start = end - intervalToMs(durations[idx]) * NS_IN_MS;
```
Comment (gabor) [inline public/app/plugins/datasource/loki/components/stats.ts:78]: i think the `end` should be the currently-selected-grafana-time-range-end, WDYT?

**Review by gabor — COMMENTED** [review]
very nice, i left a comment because i think we should always calculate the time-range based on some parts of the grafana-time-range.. but i may be wrong...

File: public/app/plugins/datasource/loki/components/stats.ts (hunk 2 above)
Comment (gwdawson) [reply public/app/plugins/datasource/loki/components/stats.ts:78]: yeah, that makes sense

**Review by gabor — APPROVED** [review]
LGTM 👍

**Comment by ivanahuckova** [comment]
> The problem with that is that in the frontend we don't have support to interpolate $__interval to the correct value. 

For stats queries, we use only the log part of the query. So it should not matter if range is `$__interval`, `$__auto` or `5m` as we are "throwing away" that part. 

image

**Comment by ivanahuckova** [comment]
Oh, sorry disregard. Still we need to fix it as we want users to always use `$__auto` which would mean that most of the users couldn't use the functionality.

**Comment by ivanahuckova** [comment]
@gabor @gwdawson have you thought about this issue? I have couple ideas in mind:
- calculate more precise stats if we don't use variables as `[range]` and less precise if it has` $__interval`/`$__auto` (here we expect users to use mostly `$__auto` in the future)
- use `rangeUtil.calculateInterval` and just mock resolution (screen width)
- don't calculate more precise range for stats request, as it is just estimation and we can't get precise `range` without a big effort (passing containerWidth to stats reuqest)

**Comment by svennergr** [comment]
Ideally we would replace template variables anyway so this will work with vars in a dashboard. If we move that the stats request to be a supplementary query we would get all the vars we need in `request.scopedVars`, wouldn't we?

WDYT?

**Comment by ivanahuckova** [comment]
We talked with @gabor and came to following conclusions/thoughts:
1. Unfortunately, without a big effort, we are not able to send precise time range range for when Grafana built in variable is used (mostly `$__interval`). This is because we don't have the interval value and we would have to calculate it. For that we would need maxDataPoints/container size.

1. Moreover, for range query, it does not matter that much if we include `[range]` info or not, as `[range]` it is quite a small compared to the whole range (e.g. there shouldn't be a giant difference if sent time range is `6h` vs `6h and 15 seconds`. So the fact that we don't include `time range + [range]` should not have a big effect on estimated size. Therefore I would propose to not do any change in here.

1. For instant query it is different, as the`[range]` value is basically the full time range that we are sending. So in this case, I would propose to:
- use the `[range]` if it has absolute value - such as `5m`
- if `$__auto` or `$__range` is use, use the value of timerange 
- if other grafana built in variable is used (e.g. `$__interval`), write that it is not possible to estimate the size. 

The template variables in dashboard should not be an issue as we can interpolate them.

**Comment by gwdawson** [comment]
@ivanahuckova as requested:

**range queries:**
`it does not matter that much if we include [range] info or not, as [range] it is quite a small`.
so now we always just use the timerange.

**instant queries:**
- [x] use the `[range]` if it has absolute value - such as `5m`
- [x] if `$__auto` or `$__range` is use, use the value of timerange
- [x] if other grafana built in variable is used (e.g. `$__interval`), write that it is not possible to estimate the size.

File: public/app/plugins/datasource/loki/components/stats.ts [hunk 3]
```diff
@@ -52,3 +57,47 @@ export function shouldUpdateStats(
 
   return true;
 }
+
+export function getTimeRange(
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/components/stats.ts:61]: I am curious why isn't `getTimeRange` and `getStats` in here and they are not data source methods?

**Review by ivanahuckova — COMMENTED** [review]
Looks really good. Left some questions/comments. The main question (why I didn't approve) is why are `getStats` and `getTimeRange` in `stats.ts` functions and why aren't they Loki data source methods?

File: public/app/plugins/datasource/loki/datasource.test.ts [hunk 4]
```diff
@@ -1424,6 +1431,122 @@ describe('applyTemplateVariables', () => {
       );
     });
   });
+
+  describe('getTimeRange', () => {
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/datasource.test.ts:1435]:
```suggestion
  describe('getStatsTimeRange', () => {
```

**Review by ivanahuckova — APPROVED** [review]
Left 1 small comment bellow, but otherwise LGTM.  Great job!

_Sorry @gabor, I accidentally re-requested your review, when I wanted to re-request mine._

**Pull request merged**